packages/itwin/tree-widget/src/components/trees/models-tree/HierarchyBasedVisibilityHandler.ts

```diff
@@ -0,0 +1,848 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { concat, concatAll, concatWith, defer, EMPTY, firstValueFrom, forkJoin, from, map, mergeMap, of, reduce } from "rxjs";
+import { assert } from "@itwin/core-bentley";
+import { PerModelCategoryVisibility } from "@itwin/core-frontend";
+import { NodeKey } from "@itwin/presentation-common";
+import { isPresentationTreeNodeItem } from "@itwin/presentation-components";
+import { Presentation } from "@itwin/presentation-frontend";
+import { reduceWhile, toSet, toVoidPromise } from "../common/Rxjs";
+import { createQueryHandler } from "./internal/QueryHandler";
+import { createVisibilityStatus } from "./internal/Tooltip";
+import { createVisibilityChangeEventListener } from "./internal/VisibilityChangeEventListener";
+import { NodeUtils } from "./NodeUtils";
+
+import type { Id64Arg, Id64Set, Id64String } from "@itwin/core-bentley";
+import type { ElementsQueryProps, IQueryHandler as IQueryHandler } from "./internal/QueryHandler";
+import type { IFilteredPresentationTreeDataProvider } from "@itwin/presentation-components";
+import type { IVisibilityHandler, VisibilityStatus } from "../VisibilityTreeEventHandler";
+import type { IVisibilityChangeEventListener } from "./internal/VisibilityChangeEventListener";
+import type { Viewport } from "@itwin/core-frontend";
+import type { NonPartialVisibilityStatus, Visibility } from "./internal/Tooltip";
+import type { ECClassGroupingNodeKey } from "@itwin/presentation-common";
+import type { TreeNodeItem } from "@itwin/components-react";
+import type { Observable, OperatorFunction } from "rxjs";
+interface GetCategoryStatusProps {
+  categoryId: Id64String;
+  modelId: Id64String | undefined;
```

Comment:
The original visibility handler also has this parameter optional. I assumed this was a valid use case, is it not, though?

Reply:
I can't think of a case when this would happen... We only show categories that have elements and all elements have a model, so if there's a category, then there should also be a model.